### PR TITLE
[FLINK-34876][transform] Support UDF functions in transform

### PR DESCRIPTION
**Pull request by yuxiqian**
This closes FLINK-34876, aims to introduce customized logic into transform expressions, with some key features:

* supports UDF classes implemented in Java and Scala
* explicit CDC pipeline type hints
* `open` & `close` lifecycle hooks
* compatible with simple Flink `ScalarFunction`

---

One may add a `UDF` block into existing pipeline definition and call them in transform expressions:

```yaml
transform:
  - projection: addone(addone(addone(addone(1))))
    # ...

pipeline:
  user-defined-function:
    - name: addone                                   # function name used in expressions
      classpath: org.example.udf.AddOneFunctionClass # fully-qualified class name
    - name: format
      classpath: org.example.udf.FormatFunctionClass
```

where `classpath` value either implements CDC's `org.apache.flink.cdc.common.udf.UserDefinedFunction` interface or extends Flink `org.apache.flink.table.functions.ScalarFunction` abstract class.

`UserDefinedFunction` is declared as follows:

```java
public interface UserDefinedFunction {

    default DataType getReturnType() { return null; }

    default void open() throws Exception {}

    default void close() throws Exception {}
}
```

A valid `UserDefinedFunction` implementation must:

* Have a nullary constructor
* Have (at least one) `eval` method that doesn't return void

A `UserDefinedFunction` implementation may:

* Override `getReturnType` method to declare its return type.
  * It will be inferred from `eval` signature if not provided.

```java
public class TypeHintFunctionClass implements UserDefinedFunction {

    @Override
    public DataType getReturnType() {
        return DataTypes.STRING();
    }

    public Object eval() {
        // Return type could not be inferred from function signature
        return "Forty-two";
    }
}
```

* Override `open` and `close` lifecycle hooks to initialize / clean-up necessary resources.
  * `props` defined in YAML will be passed in `UserDefinedFunctionContext` as an argument.
  * UDF's lifecycle is identical to `TransformDataOperator` (or `PostTransformOperator` after FLINK-35272).
  * `UserDefinedFunction` might be constructed temporarily elsewhere, so make sure resources are handled in `open` and `close` instead of constructors.

```java
public class LifecycleFunctionClass implements UserDefinedFunction {

    private Integer counter;

    public String eval() {
        return "#" + (counter++); // This returns #1, #2, #3, ...
    }

    @Override
    public void open() {
        counter = 0;
        System.out.println("[ LifeFun ] opened.");
    }

    @Override
    public void close() {
        System.out.println("[ LifeFun ] closed. Called " + counter + " times.");
    }
}
```

`eval` method overloading is supported, and runtime invocations will be dispatched to the correct one:

```java
public class TypeOfFunctionClass implements UserDefinedFunction {

    public String eval(Boolean b) {
        return "Boolean: " + b;
    }

    public String eval(Integer i) {
        return "Integer: " + i;
    }

    public String eval(Float f) {
        return "Float: " + f;
    }

    public String eval(Double d) {
        return "Double: " + d;
    }

    public String eval(String s) {
        return "String: " + s;
    }
}
```

Flink `ScalarFunction` is supported with the following limitations:

* Flink-style type hint (`TypeInformation`) will be ignored.
* `ScalarFunction` with arguments are not supported.
* `open` and `close` methods declared will be ignored.

### CODE REVIEW

**Comment by yuxiqian** [comment]
Having some difficulties getting Calcite related materials, looking for @aiwenmo's comments.

**Comment by yuxiqian** [comment]
Updated API design & POC implementation, could @aiwenmo please take a look?

**Comment by yuxiqian** [comment]
Adjusted based on previous discussions, cleaned up changes, and added UDF-related docs.

**Comment by yuxiqian** [comment]
Rebased to latest `master`, cc @ruanhang1993 

Minor concern: By adding more E2e test cases, now it takes nearly 1h 30min to finish running e2e case, which is worrying since current timeout limit is configured to 90 minutes. Maybe something like #3514 is necessary?

File: docs/content.zh/docs/core-concept/transform.md [hunk 1]
```diff
@@ -266,6 +266,75 @@ transform:
...
+    @Override
+    public void close() throws Exception {
+        // ...
+    }
+}
+```
+
+To ease the migration from Flink SQL to Flink CDC, a Flink `ScalarFunction` could also be used as a transform UDF, with some limitations:
+
+* `ScalarFunction` with parameters is not supported.
```
Comment (ruanhang1993) [inline docs/content.zh/docs/core-concept/transform.md:312]: `ScalarFunction` which has a constructor with parameters is not supported.

File: flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/parser/YamlPipelineDefinitionParser.java [hunk 2]
```diff
@@ -199,6 +211,23 @@ private RouteDef toRouteDef(JsonNode routeNode) {
...
     }
 
+    private UdfDef toUdfDef(JsonNode udfNode) {
+        String functionName =
+                checkNotNull(
+                                udfNode.get(UDF_FUNCTION_NAME_KEY),
+                                "Missing required field \"%s\" in UDF configuration",
+                                UDF_FUNCTION_NAME_KEY)
+                        .asText();
+        String classPath =
```
Comment (ruanhang1993) [inline flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/parser/YamlPipelineDefinitionParser.java:221]:
```suggestion
        String classpath =
```

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/definition/UdfDef.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cdc.composer.definition;
+
+import java.util.Objects;
+
+/**
+ * Definition of a transformation.
```
Comment (ruanhang1993) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/definition/UdfDef.java:23]:
```suggestion
 * Definition of a user-defined function.
```

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java [hunk 4]
```diff
@@ -96,6 +98,11 @@ public PipelineExecution compose(PipelineDef pipelineDef) {
         int parallelism = pipelineDef.getConfig().get(PipelineOptions.PIPELINE_PARALLELISM);
         env.getConfig().setParallelism(parallelism);
 
+        List<Tuple2<String, String>> udfFunctions =
+                pipelineDef.getUdfs().stream()
+                        .map(udf -> Tuple2.of(udf.getName(), udf.getClassPath()))
+                        .collect(Collectors.toList());
```
Comment (ruanhang1993) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java:104]: This translation should be placed in the `transformTranslator`.

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java [hunk 5]
```diff
@@ -106,7 +113,9 @@ public PipelineExecution compose(PipelineDef pipelineDef) {
 
         // Build TransformSchemaOperator for processing Schema Event
         TransformTranslator transformTranslator = new TransformTranslator();
-        stream = transformTranslator.translateSchema(stream, pipelineDef.getTransforms());
+        stream =
+                transformTranslator.translateSchema(
+                        stream, pipelineDef.getTransforms(), udfFunctions);
```
Comment (ruanhang1993) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java:118]:
```suggestion
                        stream, pipelineDef.getTransforms(), pipelineDef.getUdfs());
```

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java [hunk 6]
```diff
@@ -125,7 +134,8 @@ public PipelineExecution compose(PipelineDef pipelineDef) {
                         stream,
                         pipelineDef.getTransforms(),
                         schemaOperatorIDGenerator.generate(),
-                        pipelineDef.getConfig().get(PipelineOptions.PIPELINE_LOCAL_TIME_ZONE));
+                        pipelineDef.getConfig().get(PipelineOptions.PIPELINE_LOCAL_TIME_ZONE),
+                        udfFunctions);
```
Comment (ruanhang1993) [inline flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/FlinkPipelineComposer.java:138]:
```suggestion
                        pipelineDef.getUdfs());
```

File: flink-cdc-pipeline-udf-examples/pom.xml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                        <goals>
+                            <goal>compile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+
+        </plugins>
+    </build>
+</project>
```
Comment (ruanhang1993) [inline flink-cdc-pipeline-udf-examples/pom.xml:104]: Add an empty line.

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/UserDefinedFunctionDescriptor.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.stream.Collectors;
+
+/** Descriptor of a UDF function. */
+@Internal
+public class UserDefinedFunctionDescriptor implements Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    private final String name;
+    private final String classPath;
```
Comment (ruanhang1993) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/UserDefinedFunctionDescriptor.java:37]:
```suggestion
    private final String classpath;
```

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/udf/UserDefinedFunction.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Base interface for creating a UDF in transform projection and filtering expressions. You should
+ * define at least one {@code eval} method.
+ */
+@PublicEvolving
+public interface UserDefinedFunction {
+    default DataType getReturnType() {
+        return null;
+    }
```
Comment (ruanhang1993) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/udf/UserDefinedFunction.java:31]:
```suggestion
    default Optional<DataType> getReturnType() {
        return Optional.empty();
    }
```

Comment (yuxiqian) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/udf/UserDefinedFunction.java:31]: I have some doubt on this. This is meant to be a interface protocol, and any reasonable overwrite of this method should not return an empty value.

**Comment by yuxiqian** [comment]
Seems CI is failing due to an expired link in Doris docs. Rebased & fixed.

**Comment by yuxiqian** [comment]
Rebased with `master` and resolved conflicts. cc @leonardBang @ruanhang1993

**Review by ruanhang1993 — APPROVED** [review]

**Review by aiwenmo — APPROVED** [review]
LGTM

**Comment by mamineturki** [comment]
@yuxiqian Hi, I have been trying to run an example with this new feature but I always get `java.lang.ClassNotFoundException: org.example.udf.AddOneFunctionClass`

I tried loading the udf function using the path in the PR description, as well as the package name `org.apache.flink.cdc.udf.examples.scala.AddOneFunctionClass` with no luck.

I imported the function by downloading the pre-packaged jar `wget https://repo1.maven.org/maven2/org/apache/flink/flink-cdc-pipeline-udf-examples/3.2.0/flink-cdc-pipeline-udf-examples-3.2.0.jar`

please advise on what am I missing exactly. Thanks

**Comment by yuxiqian** [comment]
Hi @mamineturki, `flink-cdc-pipeline-udf-examples` are for testing purposes only. If you'd like to test it, using `org.apache.flink.cdc.udf.examples.scala.AddOneFunctionClass` should work. Seems it wasn't successfully loaded by Flink classloaders in your case.

Have you tried putting `.jar` file into Flink's `/lib` path (not Flink CDC `/lib`!) and restart the Flink cluster? If that doesn't work, try passing it with `--jar XXX.jar` argument when submitting pipeline job with `bin/flink-cdc.sh`.

**Comment by mamineturki** [comment]
Thanks for the speedy reply @yuxiqian . moving the jar to `Flink lib` did not help. passing the jar with `--jar` caused the `Flink-cdc.sh` command to get stuck with no logs, and no submitted job.

**Comment by mamineturki** [comment]
Thanks for the speedy response @yuxiqian . passing the jar using `--jar` worked for me.

**Comment by mamineturki** [comment]
Hi @yuxiqian I have another question please.
I was able to produce custom UDF functions that take one, two or multiple arguments. what I am looking to test next is a UDF that generates multiple columns as output (for example splitting a full name into first and last or one hot encoding of a column) 
is this currently supported? Thanks,

**Comment by yuxiqian** [comment]
Hi @mamineturki, unfortunately it's not possible now, as YAML UDF is basically equivalent to Flink's `ScalarFunction`.

(For `TableFunction` or `AggregationFunction` support, feel free to create another feature request ticket on [JIRA](https://issues.apache.org/jira/projects/FLINK/summary).)

**Pull request merged**